From 3e5d5c19409d2d65207452a47a1921a3c5ad3fdf Mon Sep 17 00:00:00 2001
From: Harshitha Prem <quic_hprem@quicinc.com>
Date: Mon, 26 Sep 2022 10:51:25 +0530
Subject: [PATCH] ath12k: [single wiphy] incorrect channel survey dump

For survey dump collection, freq_to_idx() api is used
to map the index of particular frequency to populate its
survey.

In case of single wiphy, wiphy is aware of all the
frequency bands present, the index was incorrect because
the mapping did not consider whether the freq band is
present in its radio structure (ar). Hence, channel survey
sent to userspace was invalid.

To handle this, the freq_to_idx() api should map index for the
particular band in the ar structure.

Signed-off-by: Harshitha Prem <quic_hprem@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/wmi.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -6065,6 +6065,9 @@ static int freq_to_idx(struct ath12k *ar
 	int band, ch, idx = 0;
 
 	for (band = NL80211_BAND_2GHZ; band < NUM_NL80211_BANDS; band++) {
+		if (!ar->mac.sbands[band].channels)
+			continue;
+
 		sband = hw->wiphy->bands[band];
 		if (!sband)
 			continue;
